merlin/analysis/filterbarcodes.py

```diff
@@ -57,118 +58,301 @@ def _run_analysis(self, fragmentIndex):
         barcodeDB.write_barcodes(currentBC, fov=fragmentIndex)
 
 
-class AdaptiveFilterBarcodes(analysistask.AnalysisTask):
+class GenerateAdaptiveThreshold(analysistask.AnalysisTask):
 
     """
-    An analysis task that filters barcodes based on a mean intensity threshold
-    for each area based on the abundance of blank barcodes. The threshold
-    is selected to achieve a specified misidentification rate.
+    An analysis task that generates a three-dimension mean intenisty,
+    area, minimum distance histogram for barcodes as they are decoded.
     """
 
     def __init__(self, dataSet, parameters=None, analysisName=None):
         super().__init__(dataSet, parameters, analysisName)
 
-        if 'misidentification_rate' not in self.parameters:
-            self.parameters['misidentification_rate'] = 0.05
+        # ensure decode_task is specified
+        decodeTask = self.parameters['decode_task']
 
     def fragment_count(self):
         return len(self.dataSet.get_fovs())
 
-    def get_barcode_database(self):
-        return barcodedb.PyTablesBarcodeDB(self.dataSet, self)
-
     def get_estimated_memory(self):
-        return 1000
+        return 5000
 
     def get_estimated_time(self):
-        return 30
+        return 1800
 
     def get_dependencies(self):
-        return [self.parameters['decode_task']]
+        return [self.parameters['run_after_task']]
 
-    def get_adaptive_thresholds(self):
+    def get_blank_count_histogram(self) -> np.ndarray:
+        return self.dataSet.load_numpy_analysis_result('blank_counts', self)
+
+    def get_coding_count_histogram(self) -> np.ndarray:
+        return self.dataSet.load_numpy_analysis_result('coding_counts', self)
+
+    def get_total_count_histogram(self) -> np.ndarray:
+        return self.get_blank_count_histogram() \
+               + self.get_coding_count_histogram()
+
+    def get_area_bins(self) -> np.ndarray:
+        return self.dataSet.load_numpy_analysis_result('area_bins', self)
+
+    def get_distance_bins(self) -> np.ndarray:
         return self.dataSet.load_numpy_analysis_result(
-            'adaptive_thresholds', self)
+            'distance_bins', self)
 
-    def get_codebook(self) -> Codebook:
+    def get_intensity_bins(self) -> np.ndarray:
+        return self.dataSet.load_numpy_analysis_result(
+            'intensity_bins', self, None)
+
+    def get_blank_fraction_histogram(self) -> np.ndarray:
+        blankHistogram = self.get_blank_count_histogram()
+        totalHistogram = self.get_coding_count_histogram()
+        blankFraction = blankHistogram / totalHistogram
+        blankFraction[totalHistogram == 0] = np.finfo(blankFraction.dtype).max
         decodeTask = self.dataSet.load_analysis_task(
```

Comment:
This is overwriting the decodeTask variable set on line 72.

Reply:
This isn't an instance variable. I load the decode task in __init__ just so that if a decode task isn't specified an error is raised when the analysis task is created to make debugging easier, but it isn't saved into the class.